data_model/src/numeric.rs

```diff
@@ -0,0 +1,414 @@
+        }
+    }
+}
+
+fn quoted<S: serde::Serializer>(num: &u128, ser: S) -> Result<S::Ok, S::Error> {
+    ser.serialize_str(&format!("{num}"))
+}
+
+struct NumericValueVisitor;
+
+#[derive(Deserialize)]
+#[serde(field_identifier)]
+enum Discriminants {
+    #[serde(alias = "Quantity")]
+    U32,
+    U64,
+    #[serde(alias = "BigQuantity")]
+    U128,
+    Fixed,
+}
+
+impl FromStr for NumericValue {
+    type Err = ParseNumericError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s.contains('_') {
+            if s.ends_with("_u32") {
+                Ok(NumericValue::U32(
+                    s.rsplit_once("_u32")
+                        .ok_or(ParseNumericError::Format)?
+                        .0
+                        .parse()?,
+                ))
+            } else if s.ends_with("_u64") {
+                Ok(NumericValue::U64(
+                    s.rsplit_once("_u64")
+                        .ok_or(ParseNumericError::Format)?
+                        .0
+                        .parse()?,
+                ))
+            } else if s.ends_with("_u128") {
+                Ok(NumericValue::U128(
+                    s.rsplit_once("_u128")
+                        .ok_or(ParseNumericError::Format)?
+                        .0
+                        .parse()?,
+                ))
+            } else if s.ends_with("_fx") {
+                Ok(NumericValue::Fixed(
+                    s.rsplit_once("_fx")
+                        .ok_or(ParseNumericError::Format)?
+                        .0
+                        .parse()?,
+                ))
+            } else {
+                Err(ParseNumericError::Format)
+            }
+        } else if let Ok(fixed) = Fixed::from_str(s) {
+            // This is the only unambiguous numerical type which we
+            // can safely deserialize from a string.
+            Ok(NumericValue::Fixed(fixed))
+        } else {
+            Err(ParseNumericError::Format)
+        }
+    }
+}
+
+impl<'de> serde::de::Visitor<'de> for NumericValueVisitor {
+    type Value = NumericValue;
+
+    #[inline]
+    fn expecting(&self, formatter: &mut core::fmt::Formatter) -> core::fmt::Result {
+        formatter.write_str("Either a tagged `NumericalValue` enum or an untagged double-quoted floating point number")
+    }
+
+    // This entire function is necessary to avoid stepping on the same
+    // issue that caused #3330. I could shorten the amount of
+    // boilerplate, but I'd rather it remained a sore sight until we
+    // actually replace it with a single `#[derive(Deserialize)]` for
+    // an `untagged` `StringOrNumber` enum.
+    fn visit_enum<E>(self, value: E) -> Result<NumericValue, E::Error>
+    where
+        E: serde::de::EnumAccess<'de>,
+    {
+        use serde_json::Value;
+
+        match value.variant::<Discriminants>()? {
+            (Discriminants::U32, data) => {
+                let val = data.newtype_variant::<Value>()?;
```

Comment:
We use `serde_json::Value` because we want to represent `String` and number in single enum?

Reply:
We use this so that `serde` sees the u128 literal without defining another useeless struct with a visitor. This is an extremely ugly workaround, but it could have been much worse.

Reply:
Let's leave it as is for now then